engine-types/src/types/address.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{format, String, TryFrom, H160};
+use borsh::maybestd::io;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+/// Base Eth Address type
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Address(H160);
+
+#[allow(non_snake_case, dead_code)]
+pub const fn ADDRESS(addr: H160) -> Address {
+    Address(addr)
+}
+
+impl Address {
+    /// Construct Address from H160
+    pub fn new(val: H160) -> Self {
+        Self(val)
+    }
+
+    /// Get raw H160 data
+    pub fn raw(&self) -> H160 {
+        self.0
+    }
+
+    /// Encode address to string
+    pub fn encode(&self) -> String {
+        hex::encode(self.0.as_bytes())
+    }
+
+    pub fn decode(address: String) -> Result<Address, error::AddressError> {
+        let data = hex::decode(address).map_err(|_| error::AddressError::FailedDecodeHex)?;
+        if data.len() != 20 {
+            return Err(error::AddressError::IncorrectLength);
+        }
+        let mut result = [0u8; 20];
+        result.copy_from_slice(&data);
+        Ok(Address::new(H160(result)))
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+
+    pub fn from_slice(raw_addr: &[u8]) -> Self {
+        Self::new(H160::from_slice(raw_addr))
+    }
```

Comment:
`try_from_slice` should return `Result`. But in many cases we for that message usage we don't need `Result` type

Reply:
In many places we don't need to parse a slice at all (see my comment about having a `[u8; 20]` constructor). We should not have any public functions that panic because the can be used unsafely. Please remove this `from_slice` function and in places where it is used either construct from `[u8; 20]` directly or use `try_from_slice` (whichever is appropriate).